File: realapp/lib/engine/manseryeok/lunar_solar.dart
/// Port of engine/src/engine/core/lunar-solar.ts (table lookups).
library;

import 'engine_data.dart';
import 'errors.dart';

class LunarDate {
  const LunarDate({
    required this.year,
    required this.month,
    required this.day,
    required this.isLeapMonth,
  });

  final int year;
  final int month;
  final int day;
  final bool isLeapMonth;
}

class SolarDate {
  const SolarDate({required this.year, required this.month, required this.day});

  final int year;
  final int month;
  final int day;
}

String _pad(int v) => v.toString().padLeft(2, '0');

LunarDate solarToLunar(EngineData data, int year, int month, int day) {
  final key = '$year-${_pad(month)}-${_pad(day)}';
  final match = data.solarToLunar[key];
  if (match == null) {
    throw ManseryeokException(
      ManseryeokErrorCode.dataError,
      'Unsupported solar date: $key',
    );
  }
  return LunarDate(
    year: match[0],
    month: match[1],
    day: match[2],
    isLeapMonth: match[3] != 0,
  );
}

SolarDate lunarToSolar(
  EngineData data,
  int year,
  int month,
  int day,
  bool isLeapMonth,
) {
  final key = '$year-${_pad(month)}-${_pad(day)}-${isLeapMonth ? 1 : 0}';
  final match = data.lunarToSolar[key];
  if (match == null) {
    throw ManseryeokException(
      ManseryeokErrorCode.dataError,
      'Unsupported lunar date: $key',
    );
  }
  return SolarDate(year: match[0], month: match[1], day: match[2]);
}

File: engine/src/engine/core/lunar-solar.ts
import lunarSolarData from './data/lunar-solar.generated.json';

export interface SolarDateTime {
  year: number;
  month: number;
  day: number;
  hour?: number;
  minute?: number;
  second?: number;
}

export interface LunarDateTime {
  year: number;
  month: number;
  day: number;
  isLeapMonth: boolean;
  hour?: number;
  minute?: number;
  second?: number;
}

interface LunarSolarLookup {
  solarToLunar: Record<string, [number, number, number, number]>;
  lunarToSolar: Record<string, [number, number, number]>;
}

const LOOKUP = lunarSolarData as unknown as LunarSolarLookup;

function withTimeDefaults<T extends SolarDateTime | LunarDateTime>(value: T): Required<T> {
  return {
    ...value,
    hour: value.hour ?? 0,
    minute: value.minute ?? 0,
    second: value.second ?? 0,
  } as Required<T>;
}

function pad(value: number): string {
  return String(value).padStart(2, '0');
}

function toSolarKey(year: number, month: number, day: number): string {
  return `${year}-${pad(month)}-${pad(day)}`;
}

function toLunarKey(year: number, month: number, day: number, isLeapMonth: boolean): string {
  return `${year}-${pad(month)}-${pad(day)}-${isLeapMonth ? 1 : 0}`;
}

export function solarToLunar(input: SolarDateTime): LunarDateTime {
  const resolved = withTimeDefaults(input);
  const key = toSolarKey(resolved.year, resolved.month, resolved.day);
  const match = LOOKUP.solarToLunar[key];

  if (!match) {
    throw new Error(`Unsupported solar date: ${key}`);
  }

  return {
    year: match[0],
    month: match[1],
    day: match[2],
    isLeapMonth: Boolean(match[3]),
    hour: resolved.hour,
    minute: resolved.minute,
    second: resolved.second,
  };
}

export function lunarToSolar(input: LunarDateTime): SolarDateTime {
  const resolved = withTimeDefaults(input);
  const key = toLunarKey(resolved.year, resolved.month, resolved.day, resolved.isLeapMonth);
  const match = LOOKUP.lunarToSolar[key];

  if (!match) {
    throw new Error(`Unsupported lunar date: ${key}`);
  }

  return {
    year: match[0],
    month: match[1],
    day: match[2],
    hour: resolved.hour,
    minute: resolved.minute,
    second: resolved.second,
  };
}
